feat: Add an Accelerometer to the Dashboard

Atheuz opened the pull request:
Add an Accelerometer to the Dashboard. 

Implementation notes:
Acceleration is the rate of change of the velocity of an object with respect to time. 

To calculate acceleration we can use the speed of the car, specifically`vis.FrontSpeed`, which is the speed of the car in meters per second. That var is updated every frame, so if we use the `FrontSpeed` of the current frame and the previous frame, we can get acceleration without consideration to time (which makes it pretty inaccurate): `acc = currentSpeed - previousSpeed`. 

To make it more accurate we incorporate how much time has elapsed since `previousSpeed` was saved, to do this we can use `Update(float dt)` to get this, as `dt` is the time since the last frame in milliseconds. 
We need to get the `dt` to be in seconds, which can be achieved by dividing it by 1000, this is done so we have matching units (meters per second vs seconds, instead of meters per second vs milliseconds).
We then incorporate this into the calculation: `acc = (currentSpeed - previousSpeed) / (dt/1000)`, and now we have a pretty good approximation of the acceleration of the car in meters per second per second.

The value that we end up with jitters a lot, so we want to smooth that out, which we can accomplish by saving the last 4 calculated acceleration values and getting their average.

This then is the final value that we pass into the function called `RenderAccelerometer`, where we add two rectangles with one containing positive acceleration, and the other containing negative acceleration. Further we add text to make it more clear what the actual value is for each. **NOTE: I'm pretty bad at the visual aspect of these plugins and it could be better, but it could also be worse.**

Other resources:
Here's a link to a Google Sheets file that I used to determine how to calculate the acceleration given a series of `dt` and `speed` values. https://docs.google.com/spreadsheets/d/1yfhfroxRtCYhZcg04AA1XM2N8UXmlbB-EzS0HFdr9uo/edit?usp=sharing

Picture of the accelerometer rectangles on the left-most side: 
![billede](https://user-images.githubusercontent.com/202696/131362461-fc78293d-3dee-4beb-bbb5-d5452a3bd7f5.png)

codecat commented:
Awesome! I will check out the code for this soon!

codecat commented on Source/Main.as:
```diff
@@ -1,4 +1,8 @@
 bool g_visible = true;
+float g_dt = 0;
+float g_prev_speed = 0;
+array<float> g_acc = {0, 0, 0, 0};
+int g_idx = 0;
```
I don't like these being globals. Why are they not part of `DashboardAcceleration`?

Atheuz replied:
Huh, I had an issue before where state wasn't being properly saved in `DashboardAcceleration` so I kept things global, but now that I'm rechecking I might have done something wrong when I was trying to use them as `DashboardAcceleration` locals. I'll fix this.

codecat commented on Source/Settings.as:
```diff
@@ -196,6 +208,41 @@ vec4 Setting_Gearbox_TextColor = vec4(1, 1, 1, 1);
+
+
+[Setting category="Acceleration" name="Acceleration color" color]
+vec4 Setting_Acceleration_Positive_Color = vec4(0, 0.9f, 0, 1);
```
This should probably say "Positive acceleration color"

codecat commented on Source/Settings.as:
```diff
@@ -196,6 +208,41 @@ vec4 Setting_Gearbox_TextColor = vec4(1, 1, 1, 1);
+[Setting category="Acceleration" name="Border radius" drag min=0 max=50]
+float Setting_Acceleration_BorderRadius = 5.0f;
+
+[Setting category="Acceleration" name="Max Acceleration value" drag min=0 max=200]
```
Capitalized "A" here is off, plus you could probably write out "Maximum", unless it's also written as "Max" in other Dashboard settings.

codecat commented on Source/Main.as:
```diff
@@ -53,6 +63,10 @@ void RenderInterface()
 	}
 }
 
+void Update(float dt) {
+	g_dt = dt;
+}
+
```
Formatting on this function is wrong, it should be:
```
void Update(float dt)
{
}
```
Also, probably put a comment here saying what the delta time is required for.

codecat commented on Source/Things/Acceleration.as:
```diff
@@ -0,0 +1,90 @@
+		int arr_size = 4;
+		float curr_acc = ((speed - g_prev_speed) / (g_dt/1000));
+		g_prev_speed = speed;
+		g_acc[g_idx] = curr_acc;
+		g_idx = (g_idx + 1) % arr_size;
+		float sum = 0;
+		for(int n = 0; n < arr_size; n++)
+			sum += g_acc[n];
```
You could avoid arrays when smoothing using interpolation as well, which is what I'd probably prefer in this case. For example: `value = lerp(value, newValue, 0.9f);` (or some other value than 0.9, but you can play around with it to see what works well)

Additionally, maybe a setting to enable/disable smoothing could be nice.

Atheuz replied:
I'll look into using lerp instead.

Atheuz replied:
Tested out Math::Lerp and it seems to me that just getting the average over a couple of frames is better.

codecat reviewed (changes requested):
Made some in-line comments 👍

codecat commented:
Additionally, there should be an option for padding between positive & negative bars.

Atheuz commented:
Added setting to enable/disable smoothing.
Added setting for padding between positive and negative bars.
Fixed formatting for `void Update(float dt)` function.
Fixed setting names.
Moved all globals except `g_dt` into `DashboardAcceleration`

Atheuz commented:
> 
> 
> Additionally, there should be an option for padding between positive & negative bars.
> 
> The text also doesn't really look centered to me 🤔

I have issues with the text, I don't really know how to properly center it. Maybe you can help?

Atheuz commented:
Fixed text centering. 
Tested out `Math::Lerp` and it seems to me that just getting the average over a couple of frames is better.

codecat commented on Source/Settings.as:
```diff
@@ -49,6 +49,18 @@ vec2 Setting_General_WheelsPos = vec2(0.909f, 0.9f);
 vec2 Setting_General_WheelsSize = vec2(230, 370);
+
+[Setting category="General" name="Show acceleration"]
+bool Setting_General_ShowAcceleration = true;
```
I think by default make this false, just like the wheels element. It might be weird if people upgrade and they suddenly have a new widget that they're not sure of what it does.

codecat commented on Source/Settings.as:
```diff
@@ -196,6 +208,47 @@ vec4 Setting_Gearbox_TextColor = vec4(1, 1, 1, 1);
+
+[Setting category="Acceleration" name="Enable smoothing"]
+bool Setting_Acceleration_Smoothing = true;
+
```
There should be an option to change the font, just like the gearbox and all the other widgets.

codecat commented on Source/Things/Acceleration.as:
```diff
@@ -0,0 +1,122 @@
+			float avg = sum / arr_size;
+			RenderPositiveAccelerometer(offset, size, avg);
+			RenderNegativeAccelerometer(offset, size, avg);
+		}
+		else
+		{
```
Nitpick: Change to:
```asc
} else {
```

codecat commented on Source/Things/Acceleration.as:
```diff
@@ -0,0 +1,122 @@
+		vec2 offset = vec2(0.0f, 0.0f);
+		vec2 size = m_size;
+
+		if (Setting_Acceleration_Smoothing)
+		{
```
Nitpick: Change to:
```asc
if (Setting_Acceleration_Smoothing) {
```

codecat commented on Source/Things/Acceleration.as:
```diff
@@ -0,0 +1,122 @@
+			acc[idx] = curr_acc;
+			idx = (idx + 1) % arr_size;
+			float sum = 0;
+			for(int n = 0; n < arr_size; n++)
+				sum += acc[n];
```
Nitpick: Change to:
```asc
for (int n = 0; n < arr_size; n++) {
	sum += acc[n];
}
```

codecat reviewed (changes requested):
Nice, looks a lot better now! Just a few things and I'd say it's good to merge 🎉

codecat commented:
Looks great, thank you! 🎉

ZayshaaCodes commented:
Having made this recently also, cool to see this as part of the official dashboard! I was just noticing the acceleration value it's reporting isnt correct. Looks like you just need to scale the speed by 3.6f :)

codecat commented:
Oops, I missed that in review 😔 wanna do a PR for it? 😏

Atheuz commented:
See here https://github.com/codecat/tm-dashboard/pull/27
